File: p279.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int minSquares[10000];
    int findMinSquares(int num) {
        if(minSquares[num] != 0) {
            return minSquares[num];
        } else if(num == 0) {
            return 0;
        } else if(num == 1) {
            return 1;
        } else {
            int min_count = INT_MAX;
            int curr_count = 0;
            for(int i = 1; i <= int(sqrt(num)); ++i) {
                int count = 1 + findMinSquares(num - i*i);
                if(count < min_count) {
                    min_count = count;
                }
            }
            minSquares[num] = min_count;
            return min_count;
        }
    }
    int numSquares(int n) {
        memset(minSquares, sizeof(minSquares), 0);
        return findMinSquares(n);
    }
};

File: p279.py
from math import sqrt

class Solution:
    def numSquares(self, n: int) -> int:
        if sqrt(n) == int(sqrt(n)):
            return 1
        for i in range(1, int(sqrt(n))+1):
            if sqrt(n-i*i) == int(sqrt(n-i*i)):
                return 2
        while n%4 == 0:
            n //= 4
        if n%8 == 7:
            return 4
        return 3
